File: backend/app/api/v1/release_plans.py
from typing import Optional, List
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session

from app.core.database import get_db
from app.api.deps import get_current_user, require_permissions
from app.core.security_matrix import Permission
from app.models.user import User
from app.models.control_parameters.release_plan_state import ReleasePlanState
from app.repositories.release_plan_state import ReleasePlanStateRepository
from app.repositories.release_plan import ReleasePlanRepository, ReleasePlanLineRepository

from app.schemas.release_plan import (
    ReleasePlanStateCreate,
    ReleasePlanStateUpdate,
    ReleasePlanStateResponse,
    ReleasePlanCreate,
    ReleasePlanUpdate,
    ReleasePlanResponse,
    ReleasePlanListResponse,
    PaginatedReleasePlanListResponse,
    ReleasePlanLineCreate,
    ReleasePlanLineUpdate,
    ReleasePlanLineResponse,
    BulkAddLinesRequest,
    AddLinesFromRequestResponse,
    EligibleModuleLineResponse,
    LinkModuleLinesRequest,
    LinkModuleLinesResponse,
    LinkedReleasePlanEntry,
)
from app.services.release_plan_service import ReleasePlanService

router = APIRouter(prefix="/release-plans", tags=["Release Plans"])


# ─── Control Parameter: ReleasePlanState ─────────────────────────────────────

@router.get("/states/", response_model=List[ReleasePlanStateResponse])
def list_release_plan_states(
    include_inactive: bool = Query(False),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    repo = ReleasePlanStateRepository(db)
    if include_inactive:
        return repo.get_all()
    return repo.get_active()


@router.post(
    "/states/",
    response_model=ReleasePlanStateResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_release_plan_state(
    data: ReleasePlanStateCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_permissions([Permission.SYSTEM_MANAGE])),
):
    repo = ReleasePlanStateRepository(db)
    if repo.get_by(name=data.name):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Release plan state with this name already exists",
        )
    obj = ReleasePlanState(**data.model_dump())
    return repo.create(obj)


@router.patch("/states/{state_id}", response_model=ReleasePlanStateResponse)
def update_release_plan_state(
    state_id: int,
    data: ReleasePlanStateUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_permissions([Permission.SYSTEM_MANAGE])),
):
    repo = ReleasePlanStateRepository(db)
    obj = repo.get(state_id)
    if not obj:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="State not found")
    for field, value in data.model_dump(exclude_unset=True).items():
        setattr(obj, field, value)
    return repo.update(obj)


@router.delete("/states/{state_id}", status_code=status.HTTP_204_NO_CONTENT)
def deactivate_release_plan_state(
    state_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_permissions([Permission.SYSTEM_MANAGE])),
):
    repo = ReleasePlanStateRepository(db)
    if not repo.soft_delete(state_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="State not found")


@router.post("/states/{state_id}/restore", response_model=ReleasePlanStateResponse)
def restore_release_plan_state(
    state_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_permissions([Permission.SYSTEM_MANAGE])),
):
    repo = ReleasePlanStateRepository(db)
    if not repo.restore(state_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="State not found")
    return repo.get(state_id)


# ─── Release Plans CRUD ───────────────────────────────────────────────────────

@router.get("/", response_model=PaginatedReleasePlanListResponse)
def list_release_plans(
    state_id: Optional[int] = Query(None),
    source_environment_id: Optional[int] = Query(None),
    target_environment_id: Optional[int] = Query(None),
    page: int = Query(1, ge=1),
    limit: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ReleasePlanService(db)
    items, total = service.list(
        state_id=state_id,
        source_environment_id=source_environment_id,
        target_environment_id=target_environment_id,
        page=page,
        limit=limit,
    )
    pages = (total + limit - 1) // limit if total > 0 else 0
    return PaginatedReleasePlanListResponse(
        items=items, total=total, page=page, limit=limit, pages=pages
    )


@router.post("/", status_code=status.HTTP_201_CREATED)
def create_release_plan(
    data: ReleasePlanCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ReleasePlanService(db)
    plan, warning = service.create(current_user, data)
    response = ReleasePlanResponse.model_validate(plan)
    response.permissions = service.get_permissions_payload(current_user, plan)
    result = response.model_dump()
    if warning:
        result["_warning"] = warning
    return result


@router.get("/{plan_id}", response_model=ReleasePlanResponse)
def get_release_plan(
    plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ReleasePlanService(db)
    plan = service.get(plan_id)
    response = ReleasePlanResponse.model_validate(plan)
    response.permissions = service.get_permissions_payload(current_user, plan)
    return response


@router.patch("/{plan_id}")
def update_release_plan(
    plan_id: int,
    data: ReleasePlanUpdate,
    confirm_env_change: bool = Query(
        False,
        description="Set to true to confirm clearing all lines when changing environments",
    ),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    # Check if environments are being changed and confirmation is required
    update_dict = data.model_dump(exclude_unset=True)
    if ("source_environment_id" in update_dict or "target_environment_id" in update_dict):
        repo = ReleasePlanRepository(db)
        plan = repo.get_with_relations(plan_id)
        if plan and plan.lines and not confirm_env_change:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail={
                    "message": "Changing environments will delete all existing release plan lines. "
                               "Please confirm by setting confirm_env_change=true.",
                    "requires_confirmation": True,
                    "line_count": len(plan.lines),
                },
            )

    service = ReleasePlanService(db)
    plan, warning = service.update(current_user, plan_id, data)
    response = ReleasePlanResponse.model_validate(plan)
    response.permissions = service.get_permissions_payload(current_user, plan)
    result = response.model_dump()
    if warning:
        result["_warning"] = warning
    return result


@router.delete("/{plan_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_release_plan(
    plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ReleasePlanService(db)
    service.delete(current_user, plan_id)


# ─── Release Plan Lines ───────────────────────────────────────────────────────

@router.get(
    "/{plan_id}/eligible-modules",
    response_model=List[EligibleModuleLineResponse],
)
def get_eligible_modules(
    plan_id: int,
    request_id: int = Query(..., description="Development Request ID to check eligibility for"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return all DR module lines with is_eligible + disable_reason for the Add Modules wizard."""
    service = ReleasePlanService(db)
    results = service.get_eligible_modules(plan_id, request_id)
    return [EligibleModuleLineResponse(**r) for r in results]


@router.post(
    "/{plan_id}/lines/link",
    response_model=LinkModuleLinesResponse,
    status_code=status.HTTP_201_CREATED,
)
def link_module_lines(
    plan_id: int,
    data: LinkModuleLinesRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Link selected RequestModuleLines to this Release Plan."""
    service = ReleasePlanService(db)
    added, skipped, errors = service.link_module_lines(current_user, plan_id, data.module_line_ids)
    return LinkModuleLinesResponse(added=added, skipped=skipped, errors=errors)


@router.patch("/{plan_id}/lines/{line_id}", response_model=ReleasePlanLineResponse)
def update_release_plan_line(
    plan_id: int,
    line_id: int,
    data: ReleasePlanLineUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ReleasePlanService(db)
    return service.update_line(current_user, plan_id, line_id, data)


@router.delete(
    "/{plan_id}/lines/{line_id}", status_code=status.HTTP_204_NO_CONTENT
)
def delete_release_plan_line(
    plan_id: int,
    line_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ReleasePlanService(db)
    service.delete_line(current_user, plan_id, line_id)


@router.post(
    "/{plan_id}/refresh-versions",
    response_model=ReleasePlanResponse,
)
def refresh_line_versions(
    plan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Manually trigger a refresh of all line env versions from latest sync data."""
    service = ReleasePlanService(db)
    plan = service._get_plan_or_404(plan_id)
    service._refresh_all_line_versions(plan)
    plan = service.get(plan_id)
    response = ReleasePlanResponse.model_validate(plan)
    response.permissions = service.get_permissions_payload(current_user, plan)
    return response

File: frontend/src/api/release-plans.ts
import { api } from "@/lib/api";

// ─── Types ────────────────────────────────────────────────────────────────────

export interface ReleasePlanState {
  id: number;
  name: string;
  description: string | null;
  category: string; // Open / In Progress / Closed / Failed/Cancelled
  is_active: boolean;
  display_order: number;
  created_at: string;
  updated_at: string;
}

export interface EnvironmentBrief {
  id: number;
  name: string;
  order: number;
  category: string;
}

export interface ReleasePlanStateBrief {
  id: number;
  name: string;
  category: string;
}

export interface UserBrief {
  id: number;
  username: string;
  email: string;
}

export interface DevelopmentRequestBrief {
  id: number;
  request_number: string;
  description: string;
}

export interface ReleasePlanLine {
  id: number;
  release_plan_id: number;
  request_module_line_id: number | null;
  development_request_id: number | null;
  module_id: number | null;
  module_technical_name: string | null;
  module_version: string | null;
  module_email: string | null;
  module_md5_hash: string | null;
  source_env_version: string | null;
  target_env_version: string | null;
  release_action: string | null;
  uat_ticket: string | null;
  uat_status: string | null;
  created_at: string;
  updated_at: string;
  development_request: DevelopmentRequestBrief | null;
}

export interface EligibleModuleLine {
  id: number; // RequestModuleLine.id
  module_id: number | null;
  module_technical_name: string;
  module_version: string | null;
  module_md5_sum: string | null;
  uat_status: string | null;
  uat_ticket: string | null;
  source_env_version: string | null;
  target_env_version: string | null;
  drift_action: string | null;
  is_eligible: boolean;
  disable_reason: string | null;
}

export interface LinkedReleasePlanEntry {
  release_plan_line_id: number;
  module_technical_name: string | null;
  module_version: string | null;
  plan_id: number;
  plan_number: string;
  source_env_name: string;
  target_env_name: string;
  state_name: string;
  state_category: string;
  planned_deployment_date: string | null;
  actual_deployment_date: string | null;
}

export interface LinkModuleLinesRequest {
  module_line_ids: number[];
}

export interface LinkModuleLinesResponse {
  added: ReleasePlanLine[];
  skipped: string[];
  errors: string[];
}

export interface ReleasePlan {
  id: number;
  plan_number: string;
  release_version: string;
  source_environment_id: number;
  target_environment_id: number;
  state_id: number;
  planned_deployment_date: string | null;
  actual_deployment_date: string | null;
  release_notes: string | null;
  comments: string | null;
  approved_by_id: number | null;
  deployed_by_id: number | null;
  related_release_plan_id: number | null;
  is_snapshot_taken: boolean;
  created_by_id: number | null;
  created_at: string;
  updated_at: string;
  source_environment: EnvironmentBrief;
  target_environment: EnvironmentBrief;
  state: ReleasePlanStateBrief;
  approved_by: UserBrief | null;
  deployed_by: UserBrief | null;
  created_by: UserBrief | null;
  lines: ReleasePlanLine[];
  permissions?: ReleasePlanPermissions;
  _warning?: string;
}

export interface ReleasePlanPermissions {
  can_create: boolean;
  can_modify: boolean;
  can_delete: boolean;
  can_manage_lines: boolean;
  can_transition_state: boolean;
  can_approve: boolean;
  can_deploy: boolean;
  current_role_level: number;
  current_role_name: string;
  macro_state: string;
}

export interface PaginatedReleasePlanResponse {
  items: ReleasePlan[];
  total: number;
  page: number;
  limit: number;
  pages: number;
}

// ─── Request / Input types ────────────────────────────────────────────────────

export interface ReleasePlanCreate {
  release_version?: string;
  source_environment_id: number;
  target_environment_id: number;
  state_id?: number;
  planned_deployment_date?: string;
  release_notes?: string;
  comments?: string;
  approved_by_id?: number;
  deployed_by_id?: number;
  related_release_plan_id?: number;
}

export interface ReleasePlanUpdate {
  release_version?: string;
  source_environment_id?: number;
  target_environment_id?: number;
  state_id?: number;
  planned_deployment_date?: string;
  actual_deployment_date?: string;
  release_notes?: string;
  comments?: string;
  approved_by_id?: number;
  deployed_by_id?: number;
  related_release_plan_id?: number;
}

export interface ReleasePlanStateCreate {
  name: string;
  description?: string;
  category: string;
  is_active?: boolean;
  display_order?: number;
}

export interface ReleasePlanStateUpdate {
  name?: string;
  description?: string;
  category?: string;
  display_order?: number;
}

export interface ReleasePlanLineUpdate {
  development_request_id?: number;
  module_id?: number;
  module_technical_name?: string;
  module_version?: string;
  module_email?: string;
  module_md5_hash?: string;
  uat_ticket?: string;
  uat_status?: string;
}

export interface BulkAddLinesRequest {
  development_request_id: number;
}

export interface AddLinesFromRequestResponse {
  added: ReleasePlanLine[];
  skipped: string[];
  errors: string[];
}

export interface ReleasePlanFilters {
  state_id?: number;
  source_environment_id?: number;
  target_environment_id?: number;
}

// ─── API ──────────────────────────────────────────────────────────────────────

export const releasePlansApi = {
  // States (control parameter)
  getStates: async (includeInactive = false): Promise<ReleasePlanState[]> => {
    const res = await api.get<ReleasePlanState[]>("/release-plans/states/", {
      params: { include_inactive: includeInactive },
    });
    return res.data;
  },

  createState: async (data: ReleasePlanStateCreate): Promise<ReleasePlanState> => {
    return (await api.post<ReleasePlanState>("/release-plans/states/", data)).data;
  },

  updateState: async (id: number, data: ReleasePlanStateUpdate): Promise<ReleasePlanState> => {
    return (await api.patch<ReleasePlanState>(`/release-plans/states/${id}`, data)).data;
  },

  deactivateState: async (id: number): Promise<void> => {
    await api.delete(`/release-plans/states/${id}`);
  },

  restoreState: async (id: number): Promise<ReleasePlanState> => {
    return (await api.post<ReleasePlanState>(`/release-plans/states/${id}/restore`)).data;
  },

  getStatesAll: async (): Promise<ReleasePlanState[]> => {
    const res = await api.get<ReleasePlanState[]>("/release-plans/states/", {
      params: { include_inactive: true },
    });
    return res.data;
  },

  // Plans
  list: async (
    filters?: ReleasePlanFilters,
    page = 1,
    limit = 20
  ): Promise<PaginatedReleasePlanResponse> => {
    const params = new URLSearchParams();
    params.append("page", page.toString());
    params.append("limit", limit.toString());
    if (filters?.state_id) params.append("state_id", filters.state_id.toString());
    if (filters?.source_environment_id)
      params.append("source_environment_id", filters.source_environment_id.toString());
    if (filters?.target_environment_id)
      params.append("target_environment_id", filters.target_environment_id.toString());
    return (
      await api.get<PaginatedReleasePlanResponse>(`/release-plans/?${params.toString()}`)
    ).data;
  },

  get: async (id: number): Promise<ReleasePlan> => {
    return (await api.get<ReleasePlan>(`/release-plans/${id}`)).data;
  },

  create: async (data: ReleasePlanCreate): Promise<ReleasePlan> => {
    return (await api.post<ReleasePlan>("/release-plans/", data)).data;
  },

  update: async (
    id: number,
    data: ReleasePlanUpdate,
    confirmEnvChange = false
  ): Promise<ReleasePlan> => {
    const params = confirmEnvChange ? "?confirm_env_change=true" : "";
    return (await api.patch<ReleasePlan>(`/release-plans/${id}${params}`, data)).data;
  },

  delete: async (id: number): Promise<void> => {
    await api.delete(`/release-plans/${id}`);
  },

  // Eligible modules for Add-Modules wizard
  getEligibleModules: async (planId: number, requestId: number): Promise<EligibleModuleLine[]> => {
    return (
      await api.get<EligibleModuleLine[]>(`/release-plans/${planId}/eligible-modules`, {
        params: { request_id: requestId },
      })
    ).data;
  },

  linkModuleLines: async (planId: number, data: LinkModuleLinesRequest): Promise<LinkModuleLinesResponse> => {
    return (
      await api.post<LinkModuleLinesResponse>(`/release-plans/${planId}/lines/link`, data)
    ).data;
  },

  updateLine: async (
    planId: number,
    lineId: number,
    data: ReleasePlanLineUpdate
  ): Promise<ReleasePlanLine> => {
    return (
      await api.patch<ReleasePlanLine>(`/release-plans/${planId}/lines/${lineId}`, data)
    ).data;
  },

  deleteLine: async (planId: number, lineId: number): Promise<void> => {
    await api.delete(`/release-plans/${planId}/lines/${lineId}`);
  },

  // Linked plans for a DR
  getLinkedPlansForDr: async (requestId: number): Promise<LinkedReleasePlanEntry[]> => {
    return (
      await api.get<LinkedReleasePlanEntry[]>(
        `/development-requests/requests/linked-plans/${requestId}`
      )
    ).data;
  },

  refreshVersions: async (planId: number): Promise<ReleasePlan> => {
    return (await api.post<ReleasePlan>(`/release-plans/${planId}/refresh-versions`)).data;
  },
};
